core/unit_test/tstParticleGridCommunication.hpp

```diff
@@ -0,0 +1,445 @@
+    int nr = 0;
+    for ( int k = -1; k < 2; ++k )
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i, ++nr )
+            {
+                id_host( nr ) = my_rank;
+                pos_host( nr, 0 ) = center_x + i * shift_x;
+                pos_host( nr, 1 ) = center_y + j * shift_y;
+                pos_host( nr, 2 ) = center_z + k * shift_z;
+            }
+    // Add a particle on rank zero to force some resizing for sends.
+    if ( my_rank == 0 )
+    {
+        num_data++;
+        data_host.resize( num_data );
+        auto id_host = Cabana::slice<id_index>( data_host );
+        id_host( num_data - 1 ) = my_rank;
+        auto pos_host = Cabana::slice<pos_index>( data_host );
+        pos_host( num_data - 1, 0 ) = center_x + shift_x;
+        pos_host( num_data - 1, 1 ) = center_y + shift_y;
+        pos_host( num_data - 1, 2 ) = center_z + shift_z;
+    }
+    int local_num_data = num_data;
+
+    Cabana::AoSoA<DataTypes, Kokkos::HostSpace> initial( "initial", num_data );
+    Cabana::deep_copy( initial, data_host );
+    auto pos_initial = Cabana::slice<pos_index>( initial );
+
+    // Copy to the device.
+    Cabana::AoSoA<DataTypes, TEST_MEMSPACE> data_src( "data_src", num_data );
+    Cabana::deep_copy( data_src, data_host );
+
+    if ( test_type == 0 )
+    {
+        // Do the gather with the AoSoA.
+        auto pair = gridHalo( *local_grid, data_src,
+                              std::integral_constant<std::size_t, pos_index>(),
+                              halo_width );
+        gridGather( pair.first, pair.second, data_src );
+
+        data_host.resize( data_src.size() );
+        Cabana::deep_copy( data_host, data_src );
+    }
+    // Check the results.
+    int new_num_data = data_host.size();
+    pos_host = Cabana::slice<pos_index>( data_host );
+    id_host = Cabana::slice<id_index>( data_host );
+
+    auto ghost_x = halo_width * dx;
+    auto ghost_y = halo_width * dy;
+    auto ghost_z = halo_width * dz;
+    auto hi_x = local_mesh.highCorner( Cajita::Own(), Cajita::Dim::I );
+    auto hi_y = local_mesh.highCorner( Cajita::Own(), Cajita::Dim::J );
+    auto hi_z = local_mesh.highCorner( Cajita::Own(), Cajita::Dim::K );
+
+    // Make sure owned particles haven't changed.
+    for ( int i = 0; i < local_num_data; ++i )
+    {
+        EXPECT_DOUBLE_EQ( pos_host( i, Cajita::Dim::I ),
+                          pos_initial( i, Cajita::Dim::I ) );
+        EXPECT_DOUBLE_EQ( pos_host( i, Cajita::Dim::J ),
+                          pos_initial( i, Cajita::Dim::J ) );
+        EXPECT_DOUBLE_EQ( pos_host( i, Cajita::Dim::K ),
+                          pos_initial( i, Cajita::Dim::K ) );
+    }
+    for ( int i = local_num_data; i < new_num_data; ++i )
+    {
+        bool within_x = true;
+        bool within_y = true;
+        bool within_z = true;
+        // Make sure all ghosts are in halo region in at least one direction.
+        if ( pos_host( i, Cajita::Dim::I ) < lo_x )
```

Comment:
Could you add a particle field which contains the original MPI rank and then use that to back-calculate what the position should be exactly?